### DISCUSSION TITLE
Dmd doesn't allow load global variable in asm block

### DISCUSSION THREAD

**Discussion by ntysdd** [dlang/dmd#14351] [General]
Dmd told me

> cannot directly load global variable `x` with PIC or PIE code

on x86-64, which I think is wrong.

**Comment by ntysdd** [comment]
__gshared int x;

void f()
{
        int *p = &x;
        asm
        {
                mov RAX, p;
                // lea RAX, x;
        }
}

Got inefficient code

push   rbp
mov    rbp,rsp
sub     rsp,0x10
mov    rax,QWORD PTR [rip+0x0]        # f <_D1a1fFZv+0xf>
mov    QWORD PTR [rbp-0x8],rax
mov    rax,QWORD PTR [rbp-0x8]
leave
ret

**Reply by maxhaton** [reply]
I don't think we should worry too much about dmd code quality.

We can probably add asm pattern although I'm not entirely sure how the assembler thing we have interacts with (say) the relocation tables and so on.

**Reply by ntysdd** [reply]
Isn't dmd frontend used by ldc and gdc? How do they solve this problem?

**Reply by maxhaton** [reply]
The inline assembler is not part of the frontend technically.

The dmd one is basically an ab initio assembler inside the compiler whereas LDC and GDC both translate into an internal format which is handled by the backends - GDC only supports GNU style inline ASM

**Reply by ntysdd** [reply]
Thanks.
